### PR TITLE
Depends: Update FreeType package (CVE-2020-15999)

### PR DESCRIPTION
**Pull request by mammix2**
Update FreeType package to v2.10.4 in Depends to help mitigate against recent 0 day flaw found.
See CVE-2020-15999

### CODE REVIEW

File: depends/packages/freetype.mk [hunk 1]
```diff
@@ -1,8 +1,8 @@
 package=freetype
-$(package)_version=2.7.1
+$(package)_version=2.10.4
 $(package)_download_path=https://download.savannah.gnu.org/releases/$(package)
-$(package)_file_name=$(package)-$($(package)_version).tar.bz2
-$(package)_sha256_hash=3a3bb2c4e15ffb433f2032f50a5b5a92558206822e22bfe8cbe339af4aa82f88
+$(package)_file_name=$(package)-$($(package)_version).tar.xz
+$(package)_sha256_hash=86a854d8905b19698bbc8f23b860bc104246ce4854dcea8e3b0fb21284f75784
```
Comment (luke-jr) [inline depends/packages/freetype.mk:5]: This matches what I've had on my system since 2020-10-20

**Review by luke-jr — APPROVED** [review]
utACK

Please don't close-and-make-new PRs...

**Comment by mammix2** [comment]
> utACK
> 
> Please don't close-and-make-new PRs...
No probs, understood.

**Comment by hebasto** [comment]
From CVE-2020-15999 [description](https://cve.mitre.org/cgi-bin/cvename.cgi?name=CVE-2020-15999):
> Heap buffer overflow in Freetype in Google Chrome prior to 86.0.4240.111 allowed a remote attacker to potentially exploit heap corruption via a crafted HTML page.

So I'm wondering what is the scenario to exploit that vulnerability when running `bitcoin-qt`?

@MarcoFalke 
Why 0.21.0 milestone added? Is it error prone to change dependency package at the last moment before branching-off?

**Comment by fanquake** [comment]
> Why 0.21.0 milestone added? Is it error prone to change dependency package at the last moment before branching-off?

I'm not sure, however there's not really a rush to update this just before a branch-off, and as far as I'm aware, there's no way to "use" this exploit with `bitcoin-qt`. It requires loading a specially crafted font, which has a `.png` file embedded inside it, and taking advantage of a buffer overflow thereafter. An exploit like this is obviously easier to use in a web browser, where pages can load arbitrary resources, like fonts.

Note that when we compile FreeType in depends we also [explicitly disable png support](https://github.com/bitcoin/bitcoin/blob/7e373294a5ae819099c39d9d03d1f5a311d63cfc/depends/packages/freetype.mk#L8), so for depends (release) builds we shouldn't even be compiling the affected code ([`Load_SBit_Png`](https://savannah.nongnu.org/bugs/?59308)). If you look in the Linux depends output for Freetype you should see:
```bash
Library configuration:
  external zlib: no
  bzip2:         no
  libpng:        no
```

We also currently load `libfreetype.so` at runtime, so if anything, an end-user should make sure that the version of libfreetype on their system is updated.

I'm fine for us to update FreeType post branch off as part of #19716 (I'd already included an update there, although it's currently just to 2.10.2).

**Review by kristapsk — APPROVED** [review]
utACK efea820083b341fc461fa2eb6fb95efe2bdfc61e. Package SHA256 hash matches the one I have on my computer, downloaded by OS package manager some time ago.

**Comment by laanwj** [comment]
Thanks for the investigation @fanquake .

> Note that when we compile FreeType in depends we also explicitly disable png support, so for depends (release) builds we shouldn't even be compiling the affected code (Load_SBit_Png). If you look in the Linux depends output for Freetype you should see:

OK, so we don't compile the vulnerable code at all.

> We also currently load libfreetype.so at runtime, so if anything, an end-user should make sure that the version of libfreetype on their system is updated.

Oh, right!!! this is one of the libraries that we *only* compile as an interface to a system shared library. We don't actually use any code from depends except the headers really. This would be something where TBD stubs would be better:
```
/tmp/x/bitcoin-0.20.0/bin$ readelf -d ./bitcoin-qt |grep NEEDED
 0x0000000000000001 (NEEDED)             Shared library: [libpthread.so.0]
 0x0000000000000001 (NEEDED)             Shared library: [librt.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libfontconfig.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libfreetype.so.6]    <-----------------------------------
 0x0000000000000001 (NEEDED)             Shared library: [libxcb.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libdl.so.2]
 0x0000000000000001 (NEEDED)             Shared library: [libm.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [libgcc_s.so.1]
 0x0000000000000001 (NEEDED)             Shared library: [libc.so.6]
 0x0000000000000001 (NEEDED)             Shared library: [ld-linux-x86-64.so.2]
```
So it is sufficient if users upgrade their system package. Which they really should, but not so much because of bitcoin core.

Agree with removing it from the milestone.

**Comment by luke-jr** [comment]
In that case (we don't ship the built lib at all), it may be better to leave it at an older version simply to avoid any compatibility surprises...

**Comment by mammix2** [comment]
I would say that If there are no compatibility issues flagged on the Bitcoin release processes to include 2.10.4, then that mitigates the CVE easily and should be merged in a future milestone / release, and also from end users or developers fiddling with depends compilation switches. It's just another attack vector that is simple to mitigate with the version bump.

**Comment by laanwj** [comment]
> It's just another attack vector that is simple to mitigate with the version bump.

Given the realizations above, how is this an attack vector?

**Comment by luke-jr** [comment]
I suppose it's a danger to anyone who uses `depends` themselves (ie, not via gitian)

But I'm not sure why anyone would need freetype from depends?

**Comment by fanquake** [comment]
As mentioned above, a FreeType update was already a part of #19716, which I'll begin to split up shortly. I'll add you as a co-author in the relevant commit. Going to close this PR.

**Comment by mammix2** [comment]
ok thanks.

**Pull request closed without merging**